test/core/test_task_launcher.py

```diff
@@ -74,20 +87,31 @@ def test_create_launch_expire_assignments(self):
         for assignment in launcher.assignments:
             self.assertEqual(assignment.get_status(), AssignmentState.CREATED)
 
-        launcher.launch_units("dummy-url:3000")
-
-        for unit in launcher.units:
-            self.assertEqual(unit.get_db_status(), AssignmentState.LAUNCHED)
-        for assignment in launcher.assignments:
-            self.assertEqual(assignment.get_status(), AssignmentState.LAUNCHED)
-
         launcher.expire_units()
 
         for unit in launcher.units:
             self.assertEqual(unit.get_db_status(), AssignmentState.EXPIRED)
         for assignment in launcher.assignments:
             self.assertEqual(assignment.get_status(), AssignmentState.EXPIRED)
 
+    def test_launch_assignments(self):
+        """Initialize a launcher on a task run, then create the assignments"""
+        mock_data_array = self.get_mock_assignment_data_array()
+        launcher = TaskLauncher(
+            self.db, self.task_run, mock_data_array, max_num_concurrent_units=1
```

Comment:
I still think it's worthwhile to test with `max_num_concurrent_units` as 2, and also as 4 and 5. You can put these all in a loop so long as you call `tearDown` and `setUp` at the end of each iteration.

Reply:
for some reason `tearDown` crashes with database `KeyError` in ln 256 `core/local_database.py` but works fine without calling `setUp` only.

Reply:
When does this crash occur - you'll need to call `tearDown` right before `setUp` or else the test is going to leave some excess content in the `tmp/` directory.

Reply:
weird .. the error occurs when the `setUp` call is at the beginning of the for loop, but disappears when it's called right before `tearDown()`, why would that happen?

Reply:
Unsure, but you can think of it as a lock - once `tearDown()` has been called, if you make any calls before `setUp` the resources won't exist to execute the call.